### ISSUE TITLE
Recursive Node Editors

### ISSUE THREAD

**Issue by avlec** [Fattorino/ImNodeFlow#33]
I was pleasantly surprised that node editors within nodes worked pretty well.

There are a few quirks though listed below, and possibly others. I will refer to the node editor within a node as the "inner node editor" and the editor that contains that node the "outer node editor"

- The inner node editor can render its grid and nodes outside of the outer node editor itself and the ImGui window containing it.
![oob](https://github.com/user-attachments/assets/d06c13fb-ef17-46d7-831e-b57fee9ffccd)
- Dragging from a pin in the outer node editor to the inner node editor highlights the pin dragged to. Thankfully it doesn't connect when released. It should probably not highlight indicating a connection is possible if it cannot connect. However, going the other way doesn't do any highlighting or connecting which is good.

**Comment by Fattorino** [comment]
ok wow, this is uncharted territory, I'm surprised this works at all, to be honest.
As per the nodes being rendered outside, this is because all the nodes always get drawn and then clipped by ImGui, I've tried in the past to determine the bounding box of the visible area but failed and then had no more time to dedicate to it. So that's the source of that problem.
And for the highlighting, this is a fun bug, would have to look into it, but as I said in the other issue, I wont be able to for at least one month

**Comment by avlec** [comment]
Yes I had the crazy idea and great job for you on it working so well out of the box! Sounds good if I have some free time I might look into it a bit. I've got some interesting ideas and things I'll be playing around with.

**Comment by avlec** [comment]
Infinite recursion does as expected and crashes instantly. But limited recursion (32) does not so bad but it's kinda hard to see if part of the rendering here is breaking or if the zoom-level is not rendering the grids?
![glhf](https://github.com/user-attachments/assets/0fac8cab-7c5a-468d-9110-f2caf63be68d)

**Comment by elbadcode** [comment]
this is incredible lol

It may actually fit my use case which involves namespaced objects representing d3d command queues and resource views. Most of the time you'd only open one object window at a time but occasionally they'd reference a variable from another object. Now I never have to write logic to spawn a property from another object inside each window, it seems like it will just work. Not having a shared canvas is also nice for a game overlay. Would be crazy if I could get multiviewport working and have node windows on my other monitor connecting to the game overlay lol

What would be the criteria for the bounding box fix to be doable? I'm fairly new to imgui, but using it with reshade I think I could try to handle it in a compute shader which have mem pooling capabilities. Not entirely sure I view this quirk as a bug yet

**Comment by Fattorino** [comment]
honestly, I don't even know where to start to make all of this into an actual polished feature

**Comment by elbadcode** [comment]
All good lol if it goes anywhere maybe I'll come back with a couple changes but don't sweat it
